src/Adapter/Grid/Data/Factory/CartGridDataFactory.php

```diff
@@ -0,0 +1,127 @@
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Grid\Data\Factory;
+
+use Cart;
+use Context;
+use Currency;
+use Customer;
+use PrestaShop\PrestaShop\Core\Grid\Data\Factory\GridDataFactoryInterface;
+use PrestaShop\PrestaShop\Core\Grid\Data\GridData;
+use PrestaShop\PrestaShop\Core\Grid\Record\RecordCollection;
+use PrestaShop\PrestaShop\Core\Grid\Record\RecordCollectionInterface;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * Decorates Doctrine data factory for carts
+ */
+final class CartGridDataFactory implements GridDataFactoryInterface
+{
+    /**
+     * @var GridDataFactoryInterface
+     */
+    private $doctrineDataFactory;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @param GridDataFactoryInterface $doctrineDataFactory
+     * @param TranslatorInterface $translator
+     */
+    public function __construct(
+        GridDataFactoryInterface $doctrineDataFactory,
+        TranslatorInterface $translator
+    ) {
+        $this->doctrineDataFactory = $doctrineDataFactory;
+        $this->translator = $translator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getData(SearchCriteriaInterface $searchCriteria)
+    {
+        $data = $this->doctrineDataFactory->getData($searchCriteria);
+
+        $records = $this->applyModifications($data->getRecords());
+
+        return new GridData($records, $data->getRecordsTotal(), $data->getQuery());
+    }
+
+    /**
+     * @param RecordCollectionInterface $records
+     *
+     * @return RecordCollection
+     */
+    private function applyModifications(RecordCollectionInterface $records)
+    {
+        $modifiedRecords = [];
+
+        foreach ($records as $record) {
+            switch ($record['status']) {
+                case 'abandoned_cart':
+                    $record['status'] = $this->translator->trans(
+                        'Abandoned cart',
+                        [],
+                        'Admin.Orderscustomers.Feature'
+                    );
+                    break;
+                case 'not_ordered':
+                    $record['status'] = $this->translator->trans('Not ordered', [], 'Admin.Orderscustomers.Feature');
+                    break;
+            }
+
+            if (empty($record['carrier_name'])) {
+                $record['carrier_name'] = '--';
+            }
+
+            if (empty($record['customer_name'])) {
+                $record['customer_name'] = '--';
+            }
+
+            $record['online'] = $record['id_guest'] ?
+                $this->translator->trans('Yes', [], 'Admin.Global') :
+                $this->translator->trans('No', [], 'Admin.Global');
+
+            $context = Context::getContext();
+            $context->cart = new Cart($record['id_cart']);
+            $context->currency = new Currency((int) $context->cart->id_currency);
```

Comment:
This is crazy 😮 is this how it was done in legacy page ?

Reply:
Yup, it was called for every item in list.

FYI https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminCartsController.php#L896

Reply:
Everything depends on context......

Reply:
https://github.com/PrestaShop/PrestaShop/issues/14131 will become 🔥